lightning/src/ln/functional_tests.rs

```diff
@@ -10189,3 +10226,335 @@ fn test_multi_post_event_actions() {
+		assert_eq!(added_monitors[0].0, OutPoint { txid: tx.txid(), index: 0 });
+		added_monitors.clear();
+	}
+
+	// The transaction should not have been broadcast before all channels are ready.
+	assert_eq!(nodes[0].tx_broadcaster.txn_broadcasted.lock().unwrap().len(), 0);
+
+	// Go through the funding_created and funding_signed flow with node 2.
+	nodes[2].node.handle_funding_created(&nodes[0].node.get_our_node_id(), &funding_created_2);
+	{
+		let mut added_monitors = nodes[2].chain_monitor.added_monitors.lock().unwrap();
+		assert_eq!(added_monitors.len(), 1);
+		assert_eq!(added_monitors[0].0, OutPoint { txid: tx.txid(), index: 1 });
+		added_monitors.clear();
+	}
+	expect_channel_pending_event(&nodes[2], &nodes[0].node.get_our_node_id());
+
+	let funding_signed_msg = get_event_msg!(nodes[2], MessageSendEvent::SendFundingSigned, nodes[0].node.get_our_node_id());
+	chanmon_cfgs[0].persister.set_update_ret(ChannelMonitorUpdateStatus::InProgress);
+	nodes[0].node.handle_funding_signed(&nodes[2].node.get_our_node_id(), &funding_signed_msg);
+	{
+		let mut added_monitors = nodes[0].chain_monitor.added_monitors.lock().unwrap();
+		assert_eq!(added_monitors.len(), 1);
+		assert_eq!(added_monitors[0].0, OutPoint { txid: tx.txid(), index: 1 });
+		added_monitors.clear();
+	}
+
+	// The transaction should not have been broadcast before persisting all monitors has been
+	// completed.
+	assert_eq!(nodes[0].tx_broadcaster.txn_broadcasted.lock().unwrap().len(), 0);
+	assert_eq!(nodes[0].node.get_and_clear_pending_events().len(), 0);
+
+	// Complete the persistence of the monitor.
+	nodes[0].chain_monitor.complete_sole_pending_chan_update(
+		&OutPoint { txid: tx.txid(), index: 1 }.to_channel_id()
+	);
+	let events = nodes[0].node.get_and_clear_pending_events();
+
+	// The transaction should only have been broadcast now.
+	assert_eq!(nodes[0].tx_broadcaster.txn_broadcasted.lock().unwrap().len(), 1);
+	assert_eq!(nodes[0].tx_broadcaster.txn_broadcasted.lock().unwrap()[0], tx);
+	nodes[0].tx_broadcaster.txn_broadcasted.lock().unwrap().clear();
+
+	assert_eq!(events.len(), 2);
+	assert!(events.iter().any(|e| matches!(
+		*e,
+		crate::events::Event::ChannelPending {
+			ref counterparty_node_id,
+			..
+		} if counterparty_node_id == &nodes[1].node.get_our_node_id(),
+	)));
+	assert!(events.iter().any(|e| matches!(
+		*e,
+		crate::events::Event::ChannelPending {
+			ref counterparty_node_id,
+			..
+		} if counterparty_node_id == &nodes[2].node.get_our_node_id(),
+	)));
+
+	// Ensure that batch_funding_transaction_generated will have no effect if called twice.
+	assert!(nodes[0].node.batch_funding_transaction_generated(
+		&[
+			(&temp_chan_id_1, &nodes[1].node.get_our_node_id()),
+			(&temp_chan_id_2, &nodes[2].node.get_our_node_id()),
+		],
+		tx.clone()
+	).is_err());
+	assert!(nodes[0].node.get_and_clear_pending_msg_events().is_empty());
+	check_added_monitors!(nodes[0], 0);
+}
+
+#[test]
+fn test_disconnect_in_funding_batch() {
```

Comment:
Done. This is actually an interesting case touched upon here: https://github.com/lightningdevkit/rust-lightning/pull/2486#discussion_r1297851431. Theoretically, it is safe to tolerate channel closures and disconnects for channels that already have successfully persisted monitors. However, it might then be necessary to rebroadcast commitment transactions that we currently only do at startup I think and we would spend fees unnecessarily. Therefore, for simplicity, I'll continue closing the batch here for now.